gitpod-io/gitpod
# [refresh-credential] rotate the AWS ECR credential by schedule

## Description

Add a new component `refresh-credential` to rotate the AWS ECR authorization token by schedule.

The refresh-credential component is installed once the `containerRegistry.external.url` is an AWS ECR URL.
- The private AWS ECR URL with the format [aws_account_id.dkr.ecr.region.amazonaws.com](https://docs.aws.amazon.com/AmazonECR/latest/userguide/Registries.html#:~:text=The%20URL%20for%20your%20default%20private%20registry%20is%20https%3A//aws_account_id.dkr.ecr.region.amazonaws.com.).
- The public AWS ECR URL with the format [public.ecr.aws/\<registry-alias\>](https://docs.aws.amazon.com/AmazonECR/latest/public/docker-push-ecr-image.html#:~:text=The%20public%20registry%20format%20is%20public.ecr.aws/registry_alias).

Remember to configure the
- Create a Kubernetes secret with AWS access/secret key pair.
   ```console
   aws configure

   # default path is ~/.aws/credentials
   kubectl create secret aws-iam-credential -n default --from-file=</path/to/credentialsfile>
   ```
- `containerRegistry.inCluster = false` to indicate to use external container registry.
- `containerRegistry.external.certificate` indicates the secret name the AWS ECR authorization token write to.
- `containerRegistry.external.credentials` which points to the AWS secret that with the AWS access/secret key pair.
   Example yaml manifest:
   ```yaml
   containerRegistry:
     inCluster: false
     external:
       url: 012345678969.dkr.ecr.us-west-1.amazonaws.com
       certificate:
         kind: secret
         name: aws-ecr-credential
       credentials:
         kind: secret
         name: aws-iam-credential
   ```

The refresh-credential run as Kubernetes CronJob. The AWS ECR authorization token expires every 12 hours. The CronJob runs every 6 hours. Therefore, the on-caller have time to mitigate it once the ECR authorization token rotation fails.

We can add an alert rule when the job hits the backoffLimit. The trigger criteria is to query `kube_job_status_failed{job_name=~"refresh-credential.*",reason="BackoffLimitExceeded"} > 0`.
Reference to [kube_job_status_failed](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/job-metrics.md#:~:text=kube_job_status_failed).


https://www.loom.com/share/6d4b9dedc2df4a32bea87c12e3533230

> **Note**
> Since we haven't migrate/sync all the Gitpod component container image from GCP GCR to AWS ECR. Therefore, the secret with `.dockerconfigjson` requires to have both GCP GCR and AWS ECR authorization token because the image-builder push to AWS ECR, however the blobserve and registry-facade pull from GCP GCR.

> **Note**
> Because we set the Job `restartPolicy = OnFailure`, the pod will be terminated once the max backoff retry limit reaches. So we need to rely on the logging system to troubleshoot.

> **Warning**
> Please aware the service quotas.
> - The [_private_ AWS ECR service quotas](https://docs.aws.amazon.com/general/latest/gr/ecr-public.html#ecr-public-region).
> - The [_public_ AWS ECR service quotas](https://docs.aws.amazon.com/general/latest/gr/ecr.html).

## Related Issue(s)

Fixes #12104

## How to test


Try https://g704b4152c75da6e18a1605.workspace-preview.gitpod-io-dev.com/workspaces
1. Trigger a imagebuild
2. Check the base-image and workspace-image on the AWS ECR (us-west-1)

## Release Notes

```release-note
Support AWS ECR container registry
```

## Documentation

None

## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
